### PR TITLE
Fix/update get signed url to v4

### PR DESCRIPTION
**Pull request by mjmartinson**
Generating signed urls for GCP objects will now default to the Google V4 signing process. The legacy V2 signing process is retained under the under the function get_signed_url_v2. 

### New Features
- Updated Google url signing process to V4

### CODE REVIEW

File: cirrus/google_cloud/utils.py [hunk 1]
```diff
@@ -230,6 +234,120 @@ def _get_string_to_sign(
...
+
+    datetime_now = datetime.datetime.now()
+    request_timestamp = datetime_now.strftime("%Y%m%dT%H%M%SZ")
+    datestamp = datetime_now.strftime("%Y%m%d")
+
+    cred_scope = "{}/auto/storage/goog4_request".format(datestamp)
+    credentials = "{}/{}".format(client_id, cred_scope)
+
+    canonical_headers = ""
+    ordered_headers = collections.OrderedDict(sorted(extension_headers.items()))
```
[the remark this answers is not in the document]
Comment (mjmartinson) [reply cirrus/google_cloud/utils.py:294]: I reorder the extension headers check that was below this. Based off the example program linked in Jira we do need the extension headers (they are just called headers in the example one). I pushed again so let's see if this fixes it.

[the remark this answers is not in the document]
Comment (paulineribeyre) [reply cirrus/google_cloud/utils.py:294]: right, what i mean is that we don't have a use case for providing extra headers when we call this function, so i think we can remove this parameter and simplify the logic a bit

File: cirrus/google_cloud/utils.py [hunk 2]
```diff
@@ -230,6 +238,137 @@ def _get_string_to_sign(
...
+
+    canonical_query_string = ""
+    ordered_query_parameters = collections.OrderedDict(
+        sorted(canonical_query_params.items())
+    )
+    for k, v in ordered_query_parameters.items():
+        encoded_k = quote(str(k), safe="")
+        encoded_v = quote(str(v), safe="")
+        canonical_query_string += "{}={}&".format(encoded_k, encoded_v)
+    canonical_query_string = canonical_query_string[:-1]  # remove trailing '&'
```
Comment (paulineribeyre) [inline cirrus/google_cloud/utils.py:337]: _oof,_ i know this is copied from Google's sample code, but let's use [urllib urlencode](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode) instead

File: cirrus/google_cloud/utils.py [hunk 3]
```diff
@@ -230,6 +238,137 @@ def _get_string_to_sign(
...
+        signed_headers += "{};".format(lower_k)
+    signed_headers = signed_headers[:-1]  # remove trailing ';'
+
+    if canonical_query_params is None:
+        canonical_query_params = dict()
+    canonical_query_params["x-goog-algorithm"] = "GOOG4-RSA-SHA256"
+    canonical_query_params["x-goog-credential"] = credential
+    canonical_query_params["x-goog-date"] = request_timestamp
+    canonical_query_params["x-goog-expires"] = expires
+    canonical_query_params["x-goog-signedheaders"] = signed_headers
```
[the remark this answers is not in the document]
Comment (mattreex) [reply cirrus/google_cloud/utils.py:324]: this would alter the way they are sorted and will therefore not work with requestor pays

**Comment by mattreex** [comment]
Moved to new branch

File: cirrus/google_cloud/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (paulineribeyre) [reply cirrus/google_cloud/utils.py:318]: @mattreex if the order of the headers matters, it should be explicit in the code, not implicit with the sorting. Let's add `userProject` it at the right spot after sorting the other headers, and add a comment?

**Comment by paulineribeyre** [comment]
done in #87

**Pull request closed without merging**